// rikki-tikki/lib/classes/config/APIOptions.js
// Generated by CoffeeScript 1.9.0
var APIOptions, RikkiTikkiAPI, path, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

path = require('path');

RikkiTikkiAPI = module.parent.exports;

APIOptions = (function(_super) {
  __extends(APIOptions, _super);

  function APIOptions(params) {
    var o;
    if (params == null) {
      params = {};
    }
    APIOptions.__super__.constructor.call(this, o = _.extend({
      api_basepath: RikkiTikkiAPI.API_BASEPATH,
      api_version: RikkiTikkiAPI.API_VERSION,
      api_namespace: RikkiTikkiAPI.API_NAMESPACE,
      config_filename: RikkiTikkiAPI.CONFIG_FILENAME,
      config_path: RikkiTikkiAPI.CONFIG_PATH,
      schema_path: RikkiTikkiAPI.SCHEMA_PATH,
      destructive: RikkiTikkiAPI.DESTRUCTIVE,
      wrap_schema_exports: RikkiTikkiAPI.WRAP_SCHEMA_EXPORTS,
      adapter: RikkiTikkiAPI.ADAPTER,
      debug: RikkiTikkiAPI.DEBUG
    }, params), _.keys(o));
  }

  return APIOptions;

})(RikkiTikkiAPI.base_classes.Hash);

module.exports = APIOptions;
