// rikki-tikki/lib/classes/config/ClientOptions.js
// Generated by CoffeeScript 1.9.0
var ClientOptions, RikkiTikkiAPI, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI || module.parent.exports;

ClientOptions = (function(_super) {
  __extends(ClientOptions, _super);

  function ClientOptions(params) {
    var o;
    if (params == null) {
      params = {};
    }
    ClientOptions.__super__.constructor.call(this, o = _.extend({
      host: RikkiTikkiAPI.CLIENT_HOST,
      port: RikkiTikkiAPI.CLIENT_PORT,
      api_version: RikkiTikkiAPI.CLIENT_API_VERSION,
      app_id: RikkiTikkiAPI.CLIENT_APP_ID,
      app_id_param_name: RikkiTikkiAPI.CLIENT_APP_ID_PARAM_NAME,
      api_namespace: RikkiTikkiAPI.CLIENT_API_NAMESPACE,
      rest_key: RikkiTikkiAPI.CLIENT_REST_KEY,
      rest_key_param_name: RikkiTikkiAPI.CLIENT_REST_KEY_PARAM_NAME,
      protocol: RikkiTikkiAPI.CLIENT_PROTOCOL
    }, params), _.keys(o));
  }

  return ClientOptions;

})(RikkiTikkiAPI.base_classes.Hash);

module.exports = ClientOptions;
