// rikki-tikki/lib/classes/schema_tree/SchemaTree.js
// Generated by CoffeeScript 1.9.0
var RikkiTikkiAPI, SchemaTree, Util, fs, path, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

fs = require('fs');

path = require('path');

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI || module.parent.exports;

Util = RikkiTikkiAPI.Util;

SchemaTree = (function(_super) {
  __extends(SchemaTree, _super);

  function SchemaTree() {
    return SchemaTree.__super__.constructor.apply(this, arguments);
  }

  SchemaTree.prototype.replacer = RikkiTikkiAPI.Schema.replacer;

  SchemaTree.prototype.reviver = RikkiTikkiAPI.Schema.reviver;

  SchemaTree.prototype.set = function(tree, opts, callback) {
    if (typeof opts === 'function') {
      callback = opts;
      opts = {};
    }
    this.__data = _.extend(this.__data, tree);
    return this.save(callback);
  };

  SchemaTree.prototype.unset = function(attr, callback) {
    if (this.__data.hasOwnProperty(attr)) {
      delete this.__data[attr];
    }
    return this.save(callback);
  };

  SchemaTree.prototype.load = function(callback) {
    if (this.pathExists(this.__path)) {
      return SchemaTree.__super__.load.call(this, (function(_this) {
        return function(e, data) {
          if (e != null) {
            return typeof callback === "function" ? callback(e) : void 0;
          }
          _this.__data = JSON.parse("" + (JSON.stringify(data)), _this.reviver);
          return typeof callback === "function" ? callback(null, _this.__data) : void 0;
        };
      })(this));
    } else {
      this.__data = {};
      return typeof callback === "function" ? callback(null, this.__data) : void 0;
    }
  };

  SchemaTree.prototype.create = function(path, data, callback) {
    if (data == null) {
      data = {};
    }
    return SchemaTree.__super__.create.call(this, path, JSON.stringify(data, this.replacer), callback);
  };

  return SchemaTree;

})(RikkiTikkiAPI.base_classes.AbstractLoader);

module.exports = SchemaTree;
