// rikki-tikki/lib/classes/schema_tree/SchemaTreeManager.js
// Generated by CoffeeScript 1.9.0
var AppConfig, RikkiTikkiAPI, SchemaTree, SchemaTreeManager, fs, path, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

fs = require('fs');

path = require('path');

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI || module.parent.exports;

module.exports.RikkiTikkiAPI = RikkiTikkiAPI;

SchemaTree = require('./SchemaTree');

AppConfig = require('../config/AppConfig');

SchemaTreeManager = (function(_super) {
  __extends(SchemaTreeManager, _super);

  function SchemaTreeManager() {
    this.__trees = {};
    this.__path = (new AppConfig).get('trees_path');
    if (!fs.existsSync(this.__path)) {
      fs.mkdirSync(this.__path);
    }
    (this.load = (function(_this) {
      return function() {
        var e, file, stat, _i, _len, _ref, _results;
        try {
          stat = fs.statSync(_this.__path);
        } catch (_error) {
          e = _error;
          throw new Error(e);
          return false;
        }
        if (stat != null ? stat.isDirectory() : void 0) {
          _ref = fs.readdirSync(_this.__path);
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            file = _ref[_i];
            _results.push(_this.__trees[file.split('.').shift()] = new SchemaTree("" + (fs.realpathSync(_this.__path)) + path.sep + file));
          }
          return _results;
        }
      };
    })(this))();
  }

  SchemaTreeManager.prototype.createTree = function(name, data, callback) {
    if (data == null) {
      data = {};
    }
    if (!this.__trees[name]) {
      return (this.__trees[name] = new SchemaTree).create("" + this.__path + path.sep + name + ".json", data, callback);
    } else {
      return typeof callback === "function" ? callback(new Error("SchemaTree '" + name + "' already exists")) : void 0;
    }
  };

  SchemaTreeManager.prototype.destroyTree = function(name, callback) {
    var tree;
    if ((tree = this.__trees[name]) != null) {
      return tree.destroy(callback);
    }
  };

  SchemaTreeManager.prototype.getTree = function(name, callback) {
    var tree;
    return typeof callback === "function" ? callback(null, (tree = this.__trees[name]) != null ? tree : null) : void 0;
  };

  SchemaTreeManager.prototype.listTrees = function(callback) {
    return typeof callback === "function" ? callback(null, _.keys(this.__trees)) : void 0;
  };

  SchemaTreeManager.prototype.saveTree = function(name, callback) {
    var tree;
    if ((tree = this.__trees[name]) != null) {
      return tree.save(callback);
    }
  };

  SchemaTreeManager.prototype.saveAll = function(callback) {
    var eOut;
    eOut = [];
    return _.each(this.__trees, function(v, k) {
      return v.save((function(_this) {
        return function(e) {
          if (e) {
            return eOut.push(e);
          }
        };
      })(this));
    });
  };

  if (typeof callback === "function") {
    callback(eOut.length ? eOut : null);
  }

  return SchemaTreeManager;

})(RikkiTikkiAPI.base_classes.Singleton);

module.exports = SchemaTreeManager;
