// rikki-tikki/lib/classes/connections/NativeConnection.js
// Generated by CoffeeScript 1.9.0
var DSN, EventEmitter, NativeConnection, RikkiTikkiAPI, mongodb, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

mongodb = require('mongodb');

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI;

DSN = require('mongo-dsn');

EventEmitter = require('events').EventEmitter;

_ = require('underscore')._;

NativeConnection = (function(_super) {
  __extends(NativeConnection, _super);

  function NativeConnection(args) {
    this._client = mongodb.MongoClient;
    if (args != null) {
      this.connect(args);
    }
  }

  NativeConnection.prototype.handleClose = function(evt) {
    return this.emit('close', evt);
  };

  NativeConnection.prototype.connect = function(args) {
    return this.__attemptConnection((this.__dsn = new DSN(args)).toDSN());
  };

  NativeConnection.prototype.__attemptConnection = function(string) {
    var e;
    if (this.__conn != null) {
      return;
    }
    try {
      this._client.connect(string, null, (function(_this) {
        return function(e, conn) {
          if (e != null) {
            return _this.emit('error', e);
          }
          _this.__conn = conn;
          return _this.emit('open', conn);
        };
      })(this));
    } catch (_error) {
      e = _error;
      return this.emit('error', e);
    }
    return this.emit('connected', this.__conn);
  };

  NativeConnection.prototype.getConnection = function() {
    return this.__conn;
  };

  NativeConnection.prototype.getMongoDB = function() {
    return this.getConnection();
  };

  NativeConnection.prototype.getDatabaseName = function() {
    return this.getMongoDB().databaseName;
  };

  NativeConnection.prototype.getCollectionNames = function(callback) {
    return this.__conn.collections((function(_this) {
      return function(e, res) {
        return typeof callback === "function" ? callback(e, _.compact(_.map(_.values(res), function(v) {
          if ((v.s.name.match(/\.+/)) != null) {
            return null;
          } else {
            return {
              name: v.s.namespace
            };
          }
        }))) : void 0;
      };
    })(this));
  };

  NativeConnection.prototype.isConnected = function() {
    return this.__conn != null;
  };

  NativeConnection.prototype.close = function(callback) {
    if (this.isConnected()) {
      return this.__conn.close((function(_this) {
        return function(e) {
          _this.__conn = null;
          return typeof callback === "function" ? callback(e) : void 0;
        };
      })(this));
    }
  };

  return NativeConnection;

})(EventEmitter);

NativeConnection.types = {
  ObjectId: mongodb.ObjectID,
  Binary: mongodb.Binary
};

module.exports = NativeConnection;
