// rikki-tikki/lib/classes/schema/RenderableSchema.js
// Generated by CoffeeScript 1.9.0
var RenderableSchema, RikkiTikkiAPI, Util, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI;

Util = RikkiTikkiAPI.Util;

RenderableSchema = (function(_super) {
  __extends(RenderableSchema, _super);

  RenderableSchema.prototype.toJSON = function() {
    return _.clone(this);
  };

  RenderableSchema.prototype.toString = function(spacer) {
    return JSON.stringify(this, RenderableSchema.replacer, spacer);
  };

  RenderableSchema.prototype.toSource = function() {
    var ns, schema;
    ns = ((ns = RikkiTikkiAPI.API_NAMESPACE.concat('.')) !== '.' ? ns : '');
    schema = JSON.parse(this.toString());
    delete schema.name;
    return _.template(this.__template)({
      name: RikkiTikkiAPI.Util.String.capitalize(this.name),
      schema: schema,
      ns: ns,
      api_path: RikkiTikkiAPI.SCHEMA_API_REQUIRE_PATH || ""
    });
  };

  function RenderableSchema(_at_name, obj, opts) {
    this.name = _at_name;
    if (Util.Object.isOfType(obj, RikkiTikkiAPI.Schema)) {
      _.extend(this, obj);
    } else {
      RenderableSchema.__super__.constructor.call(this, obj, opts);
    }
  }

  return RenderableSchema;

})(RikkiTikkiAPI.Schema);

RenderableSchema.prototype.__template = "// template was undefined";

RenderableSchema.replacer = function(key, value) {
  if ((value != null) && (0 > _.keys(RikkiTikkiAPI.Schema.reserved).indexOf(key))) {
    return Util.Function.toString(value);
  } else {
    return void 0;
  }
};

module.exports = RenderableSchema;
