// rikki-tikki/lib/classes/base_class/AbstractMonitor.js
// Generated by CoffeeScript 1.9.0
var AbstractMonitor, ArrayCollection, RikkiTikkiAPI, Singleton, Util, _,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

ArrayCollection = require('js-arraycollection');

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI;

Util = RikkiTikkiAPI.Util;

Singleton = module.parent.exports.Singleton;

AbstractMonitor = (function(_super) {
  __extends(AbstractMonitor, _super);

  AbstractMonitor.prototype.__exclude = [];

  AbstractMonitor.prototype.__iVal = null;

  function AbstractMonitor() {
    this.filter = __bind(this.filter, this);
    var _initialized;
    AbstractMonitor.__super__.constructor.call(this);
    this.__collection = new ArrayCollection([]);
    _initialized = false;
    this.__collection.on('collectionChanged', (function(_this) {
      return function(data) {
        var type;
        type = 'changed';
        if (!_initialized) {
          type = 'init';
        }
        return _this.emit(type, data);
      };
    })(this));
    if (RikkiTikkiAPI.Util.Env.isDevelopment()) {
      setTimeout(((function(_this) {
        return function() {
          return _this.refresh(function(e, list) {
            _initialized = true;
            return _this.startPolling(_this.__iVal);
          });
        };
      })(this)), 3);
    }
  }

  AbstractMonitor.prototype.filter = function(value) {
    var item, type, _i, _len, _ref;
    if ((type = typeof value) !== 'string') {
      throw (RikkiTikkiAPI.Util.Function.getConstructorName(this)) + ".filter expected value to be a string. Type was <" + type + ">";
    }
    _ref = this.__exclude;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      item = _ref[_i];
      if (value.match(item)) {
        return false;
      }
    }
    return true;
  };

  AbstractMonitor.prototype.refresh = function(callback) {
    throw (RikkiTikkiAPI.Util.Function.getConstructorName(this)) + ".refresh(callback) is not implemented";
  };

  AbstractMonitor.prototype.startPolling = function(interval) {
    if (interval != null) {
      this.__polling_interval = interval;
    }
    this.stopPolling();
    if (this.__polling_interval != null) {
      return this.__iVal = setInterval(((function(_this) {
        return function() {
          return _this.refresh();
        };
      })(this)), this.__polling_interval);
    }
  };

  AbstractMonitor.prototype.stopPolling = function() {
    if (this.__iVal != null) {
      return clearInterval(this.__iVal);
    }
  };

  AbstractMonitor.prototype.getNames = function() {
    return _.pluck(this.getCollection(), 'name');
  };

  AbstractMonitor.prototype.getCollection = function() {
    return this.__collection.__list;
  };

  AbstractMonitor.prototype.getItemIdx = function(name) {
    return this.getNames().lastIndexOf(name);
  };

  AbstractMonitor.prototype.itemExists = function(name) {
    return this.getItemIdx(name) > -1;
  };

  return AbstractMonitor;

})(Singleton);

module.exports = AbstractMonitor;
