// rikki-tikki/lib/classes/base_class/AbstractAdapter.js
// Generated by CoffeeScript 1.9.0
var AbstractAdapter, RikkiTikkiAPI, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI;

AbstractAdapter = (function(_super) {
  __extends(AbstractAdapter, _super);

  AbstractAdapter.prototype.required = [];

  function AbstractAdapter(_at_params) {
    this.params = _at_params != null ? _at_params : {};
    if ('AbstractAdapter' === RikkiTikkiAPI.Util.Function.getConstructorName(this)) {
      throw 'AbstractAdapter can not be directly instatiated\nhint: use a subclass instead.';
    }
    _.each(this.required, (function(_this) {
      return function(param) {
        if (!(_this.params.hasOwnProperty(param))) {
          throw Error("required param '" + param + "' was not defined in the adapter params object");
        }
      };
    })(this));
  }

  AbstractAdapter.prototype.requestHandler = function(req, res) {
    throw (RikkiTikkiAPI.Util.Function.getConstructorName(this)) + ".requestHandler(req,res) is not implemented";
  };

  AbstractAdapter.prototype.responseHandler = function(req, res) {
    throw (RikkiTikkiAPI.Util.Function.getConstructorName(this)) + ".responseHandler(req,res) is not implemented";
  };

  AbstractAdapter.prototype.addRoute = function(route, method, handler) {
    throw (RikkiTikkiAPI.Util.Function.getConstructorName(this)) + ".addRoute(route, method, handler) is not implemented";
  };

  return AbstractAdapter;

})(Object);

module.exports = AbstractAdapter;
