// rikki-tikki/lib/classes/collections/CollectionManager.js
// Generated by CoffeeScript 1.9.0
var Collection, CollectionManager, Connection, RikkiTikkiAPI, Util, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI;

Util = RikkiTikkiAPI.Util;

Connection = RikkiTikkiAPI.Connection;

Collection = require('./Collection');

_ = require('underscore')._;

CollectionManager = (function(_super) {
  __extends(CollectionManager, _super);

  CollectionManager.prototype.__cache = {};

  function CollectionManager() {
    if ((this.__conn = RikkiTikkiAPI.getConnection()) == null) {
      throw 'Database is not connected';
      process.exit(1);
    }
    if (!Util.Object.isOfType(this.__conn, Connection)) {
      throw "Connection was invalid. Expected `RikkiTikkiAPI.Connection` type was '<" + (typeof this.__conn) + ">'";
      process.exit(1);
    }
    if ((this.__db = this.__conn.getMongoDB()) == null) {
      throw Error('MongoDB Connection is broken');
      process.exit(1);
    }
  }

  CollectionManager.prototype.createCollection = function(name, opts, callback) {
    if (typeof opts === 'function') {
      if (callback == null) {
        callback = opts;
      }
      opts = {};
    }
    return Collection.create(name, opts, (function(_this) {
      return function(e, collection) {
        if (e != null) {
          return typeof callback === "function" ? callback(e, null) : void 0;
        }
        _this.__cache[name] = collection instanceof Collection ? collection : new Collection(name);
        RikkiTikkiAPI.getCollectionMonitor().refresh();
        return typeof callback === "function" ? callback(e, collection) : void 0;
      };
    })(this));
  };

  CollectionManager.prototype.dropCollection = function(name, callback) {
    return this.getCollection(name, (function(_this) {
      return function(e, collection) {
        if (e != null) {
          return typeof callback === "function" ? callback(e) : void 0;
        }
        if (collection == null) {
          return typeof callback === "function" ? callback('collection undefined') : void 0;
        }
        return collection.drop(function(e, res) {
          RikkiTikkiAPI.getCollectionMonitor().refresh();
          return typeof callback === "function" ? callback(e, collection) : void 0;
        });
      };
    })(this));
  };

  CollectionManager.prototype.listCollections = function() {
    return RikkiTikkiAPI.getCollectionMonitor().getCollection();
  };

  CollectionManager.prototype.renameCollection = function(oldName, newName, callback) {
    return this.getCollection(oldName, (function(_this) {
      return function(e, collection) {
        return collection.rename(newName, {
          dropTarget: true
        }, function(e, res) {
          RikkiTikkiAPI.getCollectionMonitor().refresh();
          return typeof callback === "function" ? callback(e, res) : void 0;
        });
      };
    })(this));
  };

  CollectionManager.prototype.getCollection = function(name, callback) {
    var collection, idx;
    if (((collection = this.__cache[name]) != null) && collection instanceof Collection) {
      return typeof callback === "function" ? callback(null, collection) : void 0;
    } else {
      if (0 <= (idx = RikkiTikkiAPI.getCollectionMonitor().getItemIdx(name))) {
        return typeof callback === "function" ? callback(null, (this.__cache[name] = new Collection(name)).getCollection()) : void 0;
      }
    }
    return typeof callback === "function" ? callback('collection not found', null) : void 0;
  };

  return CollectionManager;

})(RikkiTikkiAPI.base_classes.Singleton);

module.exports = CollectionManager;
