// rikki-tikki/lib/classes/collections/Document.js
// Generated by CoffeeScript 1.9.0
var Document, RikkiTikkiAPI, Util, _,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

_ = require('underscore')._;

RikkiTikkiAPI = module.parent.exports.RikkiTikkiAPI;

Util = RikkiTikkiAPI.Util;

Document = (function(_super) {
  __extends(Document, _super);

  Document.prototype.__data = {};

  function Document(data, opts) {
    if (opts == null) {
      opts = {};
    }
    this.errors = null;
    this.__schema = opts.schema || {
      validators: {}
    };
    if (data) {
      this.setData(data);
    }
  }

  Document.prototype.setData = function(data) {
    _.each(data, (function(_this) {
      return function(v, k) {
        if (typeof v === 'string') {
          return v = Util.String.stripNull(v);
        }
      };
    })(this));
    return this.__data = data;
  };

  Document.prototype.validate = function() {
    var k, path, v, validator, _i, _len, _ref;
    if (process.env !== 'development') {
      for (k in attrs) {
        v = attrs[k];
        if ((path = this.__schema.paths[k]) != null) {
          _ref = path.validators || [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            validator = _ref[_i];
            if ((validator[0](v)) === false) {
              return validator[1];
            }
          }
        } else {
          if (k !== this.idAttribute) {
            return this.className + " has no attribute '" + k + "'";
          }
        }
      }
    }
  };

  Document.prototype.isValid = function() {
    return this.validate() === void 0;
  };

  Document.prototype.valueOf = function() {
    return this.toJSON();
  };

  Document.prototype.toJSON = function() {
    if (this.isValid()) {
      return this.__data;
    } else {
      return null;
    }
  };

  Document.prototype.serialize = function(maxDepth) {
    var branch, serialize;
    branch = {};
    serialize = (function(_this) {
      return function(doc, parentKey, maxDepth) {
        var key, subKey, value, _results;
        _results = [];
        for (key in doc) {
          value = doc[key];
          branch[subKey = "" + parentKey + key] = Util.Object.getMongoType(value);
          if ((Util.Object.isHash(value)) && (maxDepth > 0)) {
            _results.push(serialize(value, subKey + ".", maxDepth - 1));
          } else {
            _results.push(void 0);
          }
        }
        return _results;
      };
    })(this);
    serialize(this.__data, '', maxDepth);
    return branch;
  };

  return Document;

})(Object);

module.exports = Document;
